ir/v1.cpp

```diff
@@ -141,3 +142,29 @@ const IR::Type *IR::Primitive::inferOperandType(int operand) const {
             return IR::Type::Bits::get(32); }
     return IR::Type::Unknown::get();
 }
+
+Util::Enumerator<const IR::IDeclaration*>* IR::V1Program::getDeclarations() const {
+    using ScopeVal = decltype(scope)::value_type;
+
+    std::function<const IR::Node*(const ScopeVal&)> toNode =
+      [](const ScopeVal& value) -> const IR::Node* { return value.second; };
+    std::function<const IR::IDeclaration*(const IR::Node* const&)> toIDeclaration =
+      [](const IR::Node* const& node) { return node->to<IR::IDeclaration>(); };
+
+    // Return an enumerator over all Nodes in the scope which are IDeclarations.
+    return Util::Enumerator<ScopeVal>::createEnumerator(scope.begin(), scope.end())
+              ->map(toNode)
+              ->where([](const IR::Node* node) { return node->is<IR::IDeclaration>(); })
+              ->map(toIDeclaration);
+}
+
+const IR::IDeclaration* IR::V1Program::getDeclByName(cstring name) const {
```

Comment:
It sounds like it wouldn't be too tough to deal with the IDeclaration issue, but being able to distinguish between declarations with the same name but different types does seem to be quite problematic. =\